gtsfm/densify/mvs_patchmatchnet.py

```diff
@@ -0,0 +1,251 @@
+            shuffle=False,
+            num_workers=DEFAULT_WORKERS_NUMBER,
+            drop_last=False,
+        )
+
+        model = PatchmatchNet(
+            patchmatch_interval_scale=[0.005, 0.0125, 0.025],
+            propagation_range=[6, 4, 2],
+            patchmatch_iteration=[1, 2, 2],
+            patchmatch_num_sample=[8, 8, 16],
+            propagate_neighbors=[0, 8, 16],
+            evaluate_neighbors=[9, 9, 9],
+        )
+        model = nn.DataParallel(model)
+
+        # Check if cuda devices is supported, and load the pretrained model
+        #   the pretrained checkpoint should be pre-downloaded using gtsfm/download_model_weights.sh
+        if torch.cuda.is_available():
+            model.cuda()
+            state_dict = torch.load(PATCHMATCHNET_WEIGHTS_PATH)
+        else:
+            state_dict = torch.load(PATCHMATCHNET_WEIGHTS_PATH, map_location=torch.device("cpu"))
+
+        model.load_state_dict(state_dict["model"])
+        model.eval()
+
+        depth_est_list = {}
+        confidence_est_list = {}
+        with torch.no_grad():
+            for batch_idx, sample in enumerate(loader):
+                start_time = time.time()
+
+                # Check if cuda devices is supported, and store the inference data to the target device
+                if torch.cuda.is_available():
+                    sample_device = tocuda(sample)
+                else:
+                    sample_device = copy.copy(sample)
+
+                # Inference using PatchmatchNet
+                outputs = model(
+                    sample_device["imgs"],
+                    sample_device["proj_matrices"],
+                    sample_device["depth_min"],
+                    sample_device["depth_max"],
+                )
+
+                outputs = tensor2numpy(outputs)
+                del sample_device
+
+                ids = sample["idx"]
+
+                # Save depth maps and confidence maps
+                for idx, depth_est, photometric_confidence in zip(
+                    ids, outputs["refined_depth"]["stage_0"], outputs["photometric_confidence"]
+                ):
+
+                    idx = idx.cpu().numpy().tolist()
+                    depth_est_list[idx] = depth_est.copy()
+                    confidence_est_list[idx] = photometric_confidence.copy()
+
+                logger.info(
+                    "[Densify::PatchMatchNet] Iter {}/{}, time = {:.3f}".format(
+                        batch_idx + 1, len(loader), time.time() - start_time
+                    )
+                )
+
+        # Filter inference result with thresholds
+        dense_point_cloud = self.filter_depth(
+            dataset=dataset,
+            depth_list=depth_est_list,
+            confidence_list=confidence_est_list,
+            geo_pixel_thresh=thresholds[0],
+            geo_depth_thresh=thresholds[1],
+            photo_thresh=thresholds[2],
+        )
+
+        return dense_point_cloud
+
+    def filter_depth(
```

Comment:
this function is pretty long (over 100 lines). I would like us to try to stick to 10-20 line functions where we can. Is there any intuitive way to decompose it that you can think of?

Reply:
I keep it because in the original repo the `filter_depth` method looks just like this. So I guess in this way we can easily match with the original code. And there are a lot of comments in the function, too.